corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/session_nav.js

```diff
@@ -57,30 +57,25 @@ FormplayerFrontend.module("SessionNavigate", function (SessionNavigate, Formplay
         },
         renderResponse: function (menuResponse) {
             FormplayerFrontend.request("clearForm");
-            var NextScreenCollection = Backbone.Collection.extend({});
-            var nextScreenCollection;
-            //TODO: clean up this hackiness
-            if (menuResponse.commands) {
-                nextScreenCollection = new NextScreenCollection(menuResponse.commands);
-                nextScreenCollection.type = "commands";
-            } else {
-                nextScreenCollection = new NextScreenCollection(menuResponse.entities);
-                nextScreenCollection.type = "entities";
-            }
-            nextScreenCollection.title = menuResponse.title;
-            nextScreenCollection.locales = menuResponse.locales;
-            nextScreenCollection.sessionId = menuResponse.menuSessionId;
-            nextScreenCollection.headers = menuResponse.headers;
-            nextScreenCollection.styles = menuResponse.styles;
-            nextScreenCollection.tiles = menuResponse.tiles;
-            nextScreenCollection.action = menuResponse.action;
-            nextScreenCollection.breadcrumbs = menuResponse.breadcrumbs;
-            var currentFragment = Backbone.history.getFragment();
-            var urlObject = Util.CloudcareUrl.fromJson(Util.encodedUrlToObject(currentFragment));
-            urlObject.setSessionId(nextScreenCollection.sessionId);
-            var encodedUrl = Util.objectToEncodedUrl(urlObject.toJson());
+            var currentFragment,
+                urlObject,
+                encodedUrl,
+                menuCollection;
+
+            currentFragment = Backbone.history.getFragment();
+            urlObject = Util.CloudcareUrl.fromJson(Util.encodedUrlToObject(currentFragment));
+            urlObject.setSessionId(menuResponse.menuSessionId);
+            encodedUrl = Util.objectToEncodedUrl(urlObject.toJson());
+            menuResponse.appId = urlObject.appId;
+
+            menuCollection = new FormplayerFrontend.Entities.MenuSelectCollection(
+                menuResponse,
+                { parse: true }
```

Comment:
the heart of the issue is that `renderResponse` (called on form submission) duplicated the logic in the `MenuSelectCollection.parse` [function](https://github.com/dimagi/commcare-hq/blob/br/fix-undefined-app-id/corehq/apps/cloudcare/static/cloudcare/js/formplayer/entities/menu.js#L15). at some point the logic diverged and menu navigation at the end of a form was different than regular menu navigation. 

This now creates a MenuSelectCollection and calls the parse function